### DISCUSSION TITLE
Client Hints nonce hydration error

### DISCUSSION THREAD

**Discussion by marcob896** [epicweb-dev/epic-stack#776] [Q&A] [answered]
Hi, after several hours of debugging to find the root cause after getting a React 418 error on production (fly.io), I managed to deploy on fly the dev server so the error is not minified: I finally found that ClientHintsCheck nonce prop is causing it.

How can I solve it?

This is my client-hints.tsx:
```js
/**
 * This file contains utilities for using client hints for user preference which
 * are needed by the server, but are only known by the browser.
 */
import { getHintUtils } from '@epic-web/client-hints';
import {
    clientHint as colorSchemeHint,
    subscribeToSchemeChange,
} from '@epic-web/client-hints/color-scheme';
import { clientHint as timeZoneHint } from '@epic-web/client-hints/time-zone';
import { useRevalidator } from '@remix-run/react';
import * as React from 'react';
import { useRequestInfo } from '@/utils/request-info';

const hintsUtils = getHintUtils({
    theme: colorSchemeHint,
    timeZone: timeZoneHint,
    // add other hints here
});

export const { getHints } = hintsUtils;

/**
 * @returns an object with the client hints and their values
 */
export function useHints() {
    const requestInfo = useRequestInfo();
    return requestInfo.hints;
}

/**
 * @returns inline script element that checks for client hints and sets cookies
 * if they are not set then reloads the page if any cookie was set to an
 * inaccurate value.
 */
export function ClientHintCheck({ nonce }: { nonce: string }) {
    const { revalidate } = useRevalidator();
    React.useEffect(
        () => subscribeToSchemeChange(() => revalidate()),
        [revalidate]
    );

    return (
        <script
            nonce={nonce}
            dangerouslySetInnerHTML={{
                __html: hintsUtils.getClientHintCheckScript(),
            }}
        />
    );
}
```

nonce-provider.ts:
```js
import * as React from 'react'

export const NonceContext = React.createContext<string>('')
export const NonceProvider = NonceContext.Provider
export const useNonce = () => React.useContext(NonceContext)
```

implemented in root.tsx:
```js
//...
import { useNonce } from '@/utils/nonce-provider';
//...
export default function App() {
    //...
    const nonce = useNonce();
    //...

    return (
        <html
            lang='it'
            className={theme === 'light' ? 'light' : 'dark'}
        >
            <head>
                <ClientHintCheck nonce={nonce} />
                {/* .... */}
         );
}
```

If I comment out the nonce prop on the script, I get hydration error on the dangerouslySetInnerHTML prop....
I've been fighting with this sincetwo days now and can't get my head around it..

**Comment by kentcdodds** [comment, marked as answer]
What does your entry.server.ts and server/index.ts file look like? Those are where the nonce is generated

**Reply by marcob896** [reply]
ok, I definitely missed that...thanks for pointing it out.
I did not start from the epic stack but I'm taking "slices" of it when I need them.
I still have to "reveal" remix entry points and I feel overhelmed looking at the codebase because as of today I'm just setting up a "marketing" website with just a few functionalities, no need for orm, email, and all the other goodies you provided.
Do you have a suggestion on how to integrate this nonce thing without going "all-in"?
thanks anyway for your patience and your work

**Reply by kentcdodds** [reply]
If it's just a marketing site that doesn't have user data to steal them you probably don't need a nonce 🤷‍♂️

**Reply by marcob896** [reply]
It's my personal site and I'm migrating it from next to remix while doing the restyling:
- landing/about/... pages
- mdx blog
- cal.com integration to book a consultnancy
- stripe basic flow to donate
- no user login for now, i'm just planning to add some reactions and stats for blog articles with an embedded db like turso or sqlite+litefs

I integrated the client hints lib to handle theme and locale date stuff

**Reply by kentcdodds** [reply]
You could ask ChatGPT, but I'm pretty sure you won't have need for a nonce on a site users don't login to

**Reply by marcob896** [reply]
thanks. ChatGPT suggests to use it anyway to provide several security beneifits.
I actually managed to make it working: I was missing the NonceProvider in entry.server.tsx and the helmet policy.

BUT: I now get the same hydration error:

```
Warning: Prop `nonce` did not match. Server: "null" Client: ""
script
ClientHintCheck@https://marco-braglia-website-icy-sea-5986.fly.dev/app/utils/client-hints.tsx:26:32
head
html
AppExport@https://marco-braglia-website-icy-sea-5986.fly.dev/app/root.tsx:26:16
AppWithDevTools@https://marco-braglia-website-icy-sea-5986.fly.dev/node_modules/.vite/deps/remix-development-tools_client.js?v=1ca36fc9:28946:22
RenderedRoute@https://marco-braglia-website-icy-sea-5986.fly.dev/node_modules/.vite/deps/chunk-OC6U64U7.js?v=00cb64ba:856:7
RenderErrorBoundary@https://marco-braglia-website-icy-sea-5986.fly.dev/node_modules/.vite/deps/chunk-OC6U64U7.js?v=00cb64ba:813:5
DataRoutes@https://marco-braglia-website-icy-sea-5986.fly.dev/node_modules/.vite/deps/chunk-OC6U64U7.js?v=00cb64ba:2038:7
Router@https://marco-braglia-website-icy-sea-5986.fly.dev/node_modules/.vite/deps/chunk-OC6U64U7.js?v=00cb64ba:1201:7
RouterProvider@https://marco-braglia-website-icy-sea-5986.fly.dev/node_modules/.vite/deps/chunk-OC6U64U7.js?v=00cb64ba:1855:7
RemixErrorBoundary@https://marco-braglia-website-icy-sea-5986.fly.dev/node_modules/.vite/deps/chunk-OC6U64U7.js?v=00cb64ba:6595:5
RemixBrowser@https://marco-braglia-website-icy-sea-5986.fly.dev/node_modules/.vite/deps/chunk-OC6U64U7.js?v=00cb64ba:7433:46
 ```
 
 My app is deployed in dev mode [here](https://marco-braglia-website-icy-sea-5986.fly.dev/).
 Funny enough, I get the error on Edge and Firefox browsers, I don't on Brave...

**Reply by kentcdodds** [reply]
Seems similar to this: https://github.com/epicweb-dev/epic-stack/pull/774

**Reply by marcob896** [reply]
thx. That was indeed a good catch. I updated my codebase but didn't change.


In fact, in my case hydration is because on the server the nonce is NULL, not undefined as in the PR you linked.
What is really bugging me is that on localhost I don't get the error whatever the browser, while when deployed in dev mode to fly.io I get the error on all browsers but Brave 😵



and the nonce is actually there on the http response from the server:



![image](https://github.com/epicweb-dev/epic-stack/assets/65446115/00b6d0d8-3d05-48f5-b0af-a429fe77f5a5)

**Reply by marcob896** [reply]
Ok...now I'm just banging my head against the wall: it was the DASHLANE PASSWORD MANAGER extension causing the last error, that I don't have on Brave browser.

Thanks for the help, implementing Client Hints with nonce was indeed a good trip and I learnt a lot along the way.
I had several hydration errors caused by dates non rendering correclty before the nonce drama.

🙏

**Reply by kentcdodds** [reply]
I'm glad you worked it out! Cheers :)
